### ISSUE TITLE
can't specify target table 'oc_share' for update in FROM clause

### ISSUE THREAD

**Issue by voroyam** [owncloud/core#31377]
Exeption during upgrade from 10.0.7 to 10.0.8 via command line

```

Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'DELETE FROM oc_share WHERE parent NOT IN (SELECT id FROM (SELECT id FROM oc_share ORDER BY id ASC) mysqlerr1093hack)':

SQLSTATE[HY000]: General error: 1093 You can't specify target table 'oc_share' for update in FROM clause
2018-05-07T13:21:19+00:00 Update failed
```
Upgrade was tried via updater app in web, failed with freeze. Then was tried in command line with this error.

 Database: mysql 5.7.10

Related : 

https://stackoverflow.com/a/45498

Original issue from central:

https://central.owncloud.org/t/10-0-7-upgrade-to-10-0-8-failed/13240/11?u=dmitry

**Comment by butonic** [comment]
our mysql hack no longer works with versions > 5.7.6 because the optimizer removes it. from the above entioned stackoverflow post:
> **... but watch out for the query optimiser**
> 
> However, beware that from MySQL 5.7.6 and onward, the optimiser may optimise out the subquery, and still give you the error. Luckily, the  optimizer_switch variable can be used to switch off this behaviour; although I couldn't recommend doing this as anything more than a short term fix, or for small one-off tasks.
> 
> `SET optimizer_switch = 'derived_merge=off';`

**Comment by butonic** [comment]
hm try cleaning up the db out of process with
```sql
SET optimizer_switch = 'derived_merge=off';
DELETE FROM oc_share WHERE parent NOT IN (SELECT id FROM (SELECT id FROM oc_share ORDER BY id ASC) mysqlerr1093hack)
```

afterwards the query should be empty and the upgrade should succeed.

**Comment by holger8411** [comment]
Hi butonic and thanks for dealing with my problem!

I did this:

hmegies:bin holger$ ./mysql -u root -p
Enter password: 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 9680
Server version: 5.7.10 MySQL Community Server (GPL)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> use owncloud7;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> show tables from owncloud7;
+-----------------------------------+
| Tables_in_owncloud7               |
+-----------------------------------+
| oc_account_terms                  |
| oc_accounts                       |
| oc_activity                       |
| oc_activity_mq                    |
| oc_addressbookchanges             |
| oc_addressbooks                   |
| oc_appconfig                      |
| oc_audioplayer_albums             |
| oc_audioplayer_artists            |
| oc_audioplayer_genre              |
| oc_audioplayer_playlist_tracks    |
| oc_audioplayer_playlists          |
| oc_audioplayer_stats              |
| oc_audioplayer_streams            |
| oc_audioplayer_tracks             |
| oc_authtoken                      |
| oc_bookmarks                      |
| oc_bookmarks_tags                 |
| oc_calendarchanges                |
| oc_calendarobjects                |
| oc_calendars                      |
| oc_calendarsubscriptions          |
| oc_cards                          |
| oc_cards_properties               |
| oc_comments                       |
| oc_comments_read_markers          |
| oc_contacts_cards_properties      |
| oc_credentials                    |
| oc_dav_properties                 |
| oc_dav_shares                     |
| oc_documents_invite               |
| oc_documents_member               |
| oc_documents_op                   |
| oc_documents_revisions            |
| oc_documents_session              |
| oc_external_applicable            |
| oc_external_config                |
| oc_external_mounts                |
| oc_external_options               |
| oc_federated_reshares             |
| oc_file_locks                     |
| oc_filecache                      |
| oc_files_trash                    |
| oc_group_admin                    |
| oc_group_user                     |
| oc_groups                         |
| oc_jobs                           |
| oc_ldap_group_mapping             |
| oc_ldap_group_members             |
| oc_ldap_user_mapping              |
| oc_lucene_status                  |
| oc_migrations                     |
| oc_mimetypes                      |
| oc_mounts                         |
| oc_music_album_artists            |
| oc_music_albums                   |
| oc_music_ampache_sessions         |
| oc_music_ampache_users            |
| oc_music_artists                  |
| oc_music_playlist_tracks          |
| oc_music_playlists                |
| oc_music_tracks                   |
| oc_notifications                  |
| oc_polls_comments                 |
| oc_polls_dts                      |
| oc_polls_events                   |
| oc_polls_notif                    |
| oc_polls_particip                 |
| oc_polls_particip_text            |
| oc_polls_txts                     |
| oc_preferences                    |
| oc_privatedata                    |
| oc_properties                     |
| oc_schedulingobjects              |
| oc_sensorlogger_data_types        |
| oc_sensorlogger_device_data_types |
| oc_sensorlogger_device_groups     |
| oc_sensorlogger_device_types      |
| oc_sensorlogger_devices           |
| oc_sensorlogger_logs              |
| oc_share                          |
| oc_share_external                 |
| oc_storages                       |
| oc_systemtag                      |
| oc_systemtag_group                |
| oc_systemtag_object_mapping       |
| oc_trusted_servers                |
| oc_users                          |
| oc_users_external                 |
| oc_vcategory                      |
| oc_vcategory_to_object            |
+-----------------------------------+
91 rows in set (0,00 sec)

mysql> SET optimizer_switch = 'derived_merge=off';
Query OK, 0 rows affected (0,00 sec)

mysql> DELETE FROM `oc_share` WHERE `parent` NOT IN (SELECT `id` FROM (SELECT `id` FROM `oc_share` ORDER BY `id` ASC) mysqlerr1093hack);
Query OK, 0 rows affected (0,00 sec)

mysql> exit
Bye


Then I had another go at the upgrade:

hmegies:bin holger$ sudo -u _www /Library/Server/Web/Data/Sites/owncloud/occ upgrade -vvv
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-05-07T19:07:57+00:00 Set log level to debug
2018-05-07T19:07:57+00:00 Repair step: Repair MySQL database engine
2018-05-07T19:07:57+00:00 Repair step: Repair MySQL collation
2018-05-07T19:07:57+00:00 Repair info: All tables already have the correct collation -> nothing to do
2018-05-07T19:07:57+00:00 Repair step: Repair SQLite autoincrement
2018-05-07T19:07:57+00:00 Repair step: Repair orphaned reshare
2018-05-07T19:07:57+00:00 Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'DELETE FROM `oc_share` WHERE `parent` NOT IN (SELECT `id` FROM (SELECT `id` FROM `oc_share` ORDER BY `id` ASC) mysqlerr1093hack)':

SQLSTATE[HY000]: General error: 1093 You can't specify target table 'oc_share' for update in FROM clause
2018-05-07T19:07:57+00:00 Update failed
2018-05-07T19:07:57+00:00 Maintenance mode is kept active
2018-05-07T19:07:57+00:00 Reset log level

...no success though :-(

My setup:
- MacMini running OSX 10.9 and Server 2
- PHP 5.6.31
- MySQL 5.7.10
- OwnCloud 10.0.7

Thanks a lot in advance for any clue.

**Comment by holger8411** [comment]
Thank you so much dmitry and butonic! I'm up and running again (on oc 10.0.7 which I had tried to upgrade).

I had backuped my blotched upgrade and tried to manually upgrade following this tutorial (power user ugrade): https://doc.owncloud.org/server/10.0/admin_manual/maintenance/manual_upgrade.html

This did not work returning the above mentioned error. Then I tried running the mysql commands on the *new* installation. This did not work. The error remained (see above).

Then I ran the mysql commands on the old installation that had failed to upgrade: success!

Now I'm back on 10.0.7 again.

Thank you so much!
